pkg/cert-rotation/certificate-generator.go

```diff
@@ -0,0 +1,292 @@
+
+type CertGenerator struct {
+	// WebhookNames represents the names of the webhooks in the webhook server (i.e. controller-manager-nebula-operator-webhook, autoscaler-nebula-operator-webhook)
+	WebhookNames *[]string
+
+	// WebhookServerName represents the name of the webhook server associated with the certificate.
+	WebhookServerName string
+
+	// WebhookNamespace represents the namespace of the webhook server associated with the certificate.
+	WebhookNamespace string
+
+	// CertDir represents the directory to save the certificates in
+	CertDir string
+
+	// CertValidity represents the number of days the certificate should be valid for
+	CertValidity int64
+
+	// SecretName represents the name of the secret used to store the webhook certificates
+	SecretName string
+
+	// SecretNamespace represents the namespace of the secret used to store the webhook certificates
+	SecretNamespace string
+
+	// KubernetesDomain represents the custom kubernetes domain needed in the certificate
+	KubernetesDomain string
+}
+
+func (c *CertGenerator) Run() error {
+	klog.Info("Getting kubernetes configs")
+	cfg, err := ctrlruntime.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		klog.Errorf("Error building Kubernetes clientset: %v", err.Error())
+		return err
+	}
+
+	certValidityDuration := (time.Duration(c.CertValidity) * 24 * 60 * time.Minute)
+
+	// Check certificate validity
+	klog.Infof("Checking certificate validity for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+	currCertValidity, hasCert := c.getCertValidity("tls.cert")
+
+	// Initialize certificate if needed
+	if !hasCert {
+		klog.Infof("No certificate detected. Creating certificate for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+		err := c.doCertRotation(clientset, certValidityDuration)
+		if err != nil {
+			klog.Errorf("Error rotating certificate for webhook [%v/%v]: %v", c.WebhookNamespace, c.WebhookServerName, err)
+			os.Exit(1)
+		}
+		klog.Infof("Certificate created successfully for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+	} else if currCertValidity <= 2*time.Minute {
+		klog.Infof("Certificate is within 2 min of expiration. Rotating certificate for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+		err := c.doCertRotation(clientset, currCertValidity+(certValidityDuration))
+		if err != nil {
+			klog.Errorf("Error rotating certificate for webhook [%v/%v]: %v", c.WebhookNamespace, c.WebhookServerName, err)
+			os.Exit(1)
+		}
+		klog.Infof("Certificate rotated successfully for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+	}
+
+	// Start background job for rotation
+	rotateJob := cron.New()
+	rotateJob.AddFunc(fmt.Sprintf("@every %v", certValidityDuration-1*time.Minute), func() {
+		klog.Infof("Rotating certificate for webhook [%v/%v]", c.WebhookNamespace, c.WebhookServerName)
+		err := c.doCertRotation(clientset, certValidityDuration+1*time.Minute)
```

Comment:
what's the meanning

Reply:
@MegaByte875 This tells the cronjob to run the rotate function exactly 1 minute before the old certificate expires, so it generates a new certificate and updates eveything before the expiry of the old one.